### ISSUE TITLE
Free-Standing Cucumber OSM Extract Generation

### ISSUE THREAD

**Issue by daniel-j-h** [Project-OSRM/osrm-backend#3030]
At the https://wiki.openstreetmap.org/wiki/Elbe-Labe-Meeting meetup in Dresden I was asked if there's a way to extract our Cucumber OSM extract generation logic into a (potentially free-standing) project.

It seems like many people in the OSM community have to write OSM extracts by hand (e.g. @joto for testing purpose) and would love to use our `node map` + `ways` => `.osm.xml` pipeline.

Is there an easy way for us to do this? We would have to abstract over the non-osrm specific part of the support code from what I can see.

@emiltin any ideas here? I think the main issue here would be using one half of Cucumber (a testing framework) for something it wasn't designed to do. On the other hand, if this is easy to pull of, benefits many people and someone wants to do this, why not.

**Comment by emiltin** [comment]
Nice idea! They're not using cucumber, just want a command line tool to convert node maps to .osm.xml when needed?

Maybe the tool could use the Gherkin parser (which Cucumber relies on) to read and parse the feature file with it's docstrings and tables, and pass the input to a lib that converts to .osm.xml. That lib could then be reused on the OSRM test framework?

**Comment by daniel-j-h** [comment]
> Nice idea! They're not using cucumber, just want a command line tool to convert node maps to .osm.xml when needed?

Exactly: a tool parsing grid size, node map, way tags generating valid .osm.xml from it ([example](https://github.com/Project-OSRM/osrm-backend/blob/dd512812583e2105c80721d06c1a05528284a206/features/guidance/anticipate-lanes.feature#L110-L127)).

We have to `npm install` for the tests anyway, so a separate library on npm is not a blocker.

**Comment by daniel-j-h** [comment]
For the record, use-cases for this pop up again and again every now and then. Today I wrote [this feature](https://github.com/mapbox/rewrite-exit-destination-signage/blob/5cfb6494516e0c9fbef8c4f76f10e55a3841013c/tests/rewrite.feature) for a small osm-based tool instead of having to write e.g. [this and other xml files](https://github.com/mapbox/rewrite-exit-destination-signage/blob/5cfb6494516e0c9fbef8c4f76f10e55a3841013c/tests/negative/55_not_a_motorway_junction_node.osm) by hand.
